src/examples/source-3-non-det/queens.js

```diff
@@ -0,0 +1,161 @@
+/*
+  SICP JS Exercise 4.35
+  This file contains two solutions for the n-queens puzzle using non-determinism.
+  Each of them makes use of the generate-and-test paradigm.
+  The first (queens_slow) uses non-determinism for generating both the row and column
+  for each position. It does so in an optimized manner, testing the corresponding condition
+  as soon as the choice is generated, thereby reducing the search space. However, this is not
+  enough to yield a quick solution when N = 8.
+  This solution also gives repeated solutions (i.e different permutations that have all the same positions) upon backtracking.
+  The second (queens_fast) uses non-determinism in picking only the row and not the column of each position, with the
+  columns being fixed. This further reduces the search space, yielding a quick solution when N = 8.
+*/
+
+const N = 8; // the number of queens and the size of the board
+const empty_positions = null;
+
+/******************************************************************************/
+/* Slow version which uses non-determinism for both the columns and rows,     */
+/* perform testing of a required condition as soon as the choice is generated */
+/******************************************************************************/
+
+// const result_queens_slow = queens_slow(N, N);
+// pretty_print(result_queens_slow, N);
+// generate more solutions by entering 'try again' in the REPL
+
+function queens_slow(board_size, num_queens) {
+  require(num_queens <= board_size);
+  const possible_positions = enum_list(1, board_size);
+
+  const result = accumulate(
```

Comment:
Is there a reason to use accumulate, other than the fact that we want to perform the process `num_queens` times?
If not, can we use `for_each` ?

Reply:
Yes, accumulate gives us the positions that we have already, with the variable `so_far`. It is used in checking whether potential new rows and columns can be added.